test/IPSpec.hs

```diff
@@ -42,6 +45,18 @@ arbitraryIIPv6Str adrGen msklen = toIv6 <$> adrGen <*> lenGen
     toIv6 adr len = Iv6 $ show adr ++ "/" ++ show len
     lenGen = oneof [choose (minBound, -1), choose (msklen + 1, maxBound)]
 
+data PaddedIPv4 = PaddedIPv4 [Int] [Int]
+
+instance Show PaddedIPv4 where
+  show (PaddedIPv4 pads digs) = intercalate "." $ zipWith (\p d -> replicate p '0' <> show d) pads digs
+
+instance Arbitrary PaddedIPv4 where
+  arbitrary = PaddedIPv4 <$> replicateM 4 (choose (0, 4)) <*> replicateM 4 (choose (0, 255))
+  shrink (PaddedIPv4 pads digs) = (PaddedIPv4 <$> shrinkQuad pads <*> pure digs) <> (PaddedIPv4 pads <$> shrinkQuad digs)
+    where
+      shrinkQuad [0, 0, 0, 0] = []
+      shrinkQuad xs = traverse (\d -> case d of 0 -> [0]; x -> shrink x) xs
```

Comment:
I'm using the normal integer shrink here which shrinks `0` to `[]`.  I want to retain `0` and only `[]` if all are zeros.